pattern/constraint/value/AssignmentConstraint.java

```diff
@@ -0,0 +1,624 @@
+        }
+
+        public boolean isConstant() {
+            return false;
+        }
+
+        public Operation asOperation() {
+            throw TypeDBException.of(INVALID_CASTING, className(this.getClass()), className(Operation.class));
+        }
+
+        public Function asFunction() {
+            throw TypeDBException.of(INVALID_CASTING, className(this.getClass()), className(Function.class));
+        }
+
+        public ThingVar asThingVar() {
+            throw TypeDBException.of(INVALID_CASTING, className(this.getClass()), className(ThingVar.class));
+        }
+
+        public ValueVar asValueVar() {
+            throw TypeDBException.of(INVALID_CASTING, className(this.getClass()), className(ValueVar.class));
+        }
+
+        public Constant<?> asConstant() {
+            throw TypeDBException.of(INVALID_CASTING, className(this.getClass()), className(Constant.class));
+        }
+
+        @Override
+        public abstract String toString();
+
+        @Override
+        public abstract boolean equals(Object o);
+
+        @Override
+        public abstract int hashCode();
+
+        public abstract Expression clone(Map<Identifier, Variable> argMap);
+
+        public static class Operation extends Expression {
+
+            private final Op op;
+            private final Expression a;
+            private final Expression b;
+            private final int hash;
+
+            public Operation(Op op, Expression a, Expression b) {
+                this.op = op;
+                this.a = a;
+                this.b = b;
+                this.hash = Objects.hash(op, a, b);
+            }
+
+            public Op operator() {
+                return op;
+            }
+
+            public Pair<Expression, Expression> operands() {
+                return new Pair<>(a, b);
+            }
+
+            @Override
+            public boolean isOperation() {
+                return true;
+            }
+
+            @Override
+            public Operation asOperation() {
+                return this;
+            }
+
+            @Override
+            public String toString() {
+                return a.toString() + " " + op.symbol() + " " + b.toString();
+            }
+
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) return true;
+                if (o == null || getClass() != o.getClass()) return false;
+                Operation that = (Operation) o;
+                return (this.op.equals(that.op) && this.a.equals(that.a) && this.b.equals(that.b));
+            }
+
+            @Override
+            public int hashCode() {
+                return hash;
+            }
+
+            @Override
+            public Expression clone(Map<Identifier, Variable> argMap) {
+                return new Operation(op, a.clone(argMap), b.clone(argMap));
+            }
+        }
+
+        public static class Function extends Expression {
+            private final String symbol;
```

Comment:
Hmm I think this should also use  TypeQL symbol if possible? What do you think... I left a comment there that we're missing `TypeQLTokens` for Functions